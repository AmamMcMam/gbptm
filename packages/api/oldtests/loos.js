/* global describe, before, after, it */
var config = require('../../config/config');
var baseUrl = 'http://localhost:' + config.app.port;
var co = require('co');
var supertest = require('supertest');
var request = supertest(baseUrl);
var Loo = require('../../models/loo');
var loader = require('../loader.js').dataLoader;

describe('Find loos within a box (/loos/in)', function() {
  before(function(done) {
    loader(Loo, 'looBox', function(err, result) {
      done();
      return result;
    });
  });
  after(function(done) {
    co(function*() {
      yield Loo.remove({});
    }).then(done);
  });

  it('/loos/in/:sw/:ne/:nw/:se', function(done) {
    request
      .get('/loos/in/-24.2,44.5/20.3,60.4/-24.2,60.4/20.3,44.5')
      .set('Accept', 'application/json')
      .expect(200)
      .expect(function(res) {
        if (!(res.body.features.length === 12)) {
          return 'Not enough Loos';
        }
      })
      .end(done);
  });
});

describe('Find loos within radius (/loos/near)', function() {
  // Bring up a server before testing
  before(function(done) {
    loader(Loo, 'looRadius', function(err, result) {
      done();
      return result;
    });
  });
  // tear it down after
  after(function(done) {
    co(function*() {
      yield Loo.remove({});
    }).then(done);
  });

  it('/loos/near/:lon/:lat JSON with default limit', function(done) {
    request
      .get('/loos/near/-0.2068223/51.518342/')
      .set('Accept', 'application/json')
      .query({ radius: 2000 })
      .expect(200)
      .expect(function(res) {
        if (!(res.body.features.length === 5)) {
          return 'Not correct amount of loos';
        }
      })
      .end(done);
  });

  it('/loos/near/:lon/:lat JSON with limit of 17', function(done) {
    request
      .get('/loos/near/-0.2068223/51.518342/?radius=2000&limit=20')
      .set('Accept', 'application/json')
      .expect(200)
      .expect(function(res) {
        if (!(res.body.features.length === 20)) {
          return 'Not correct amount of loos';
        }
      })
      .end(done);
  });
});

describe('Find loos via ID ( /loos/:id/)', function() {
  var looGlobal = null;
  before(function(done) {
    loader(Loo, 'looID', function(err, result) {
      looGlobal = result.ops[0];
      done();
      return result;
    });
  });

  // tear it dewn after
  after(function(done) {
    co(function*() {
      yield Loo.remove({});
    }).then(done);
  });

  it('/loos/:id JSON', function(done) {
    request
      .get('/loos/' + looGlobal._id.toString())
      .set('Accept', 'application/json')
      .expect(200)
      .expect(function(res) {
        if (!res.body) {
          return 'Not enough Loos';
        }
      })
      .end(done);
  });
});
