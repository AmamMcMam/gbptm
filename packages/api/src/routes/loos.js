const express = require('express');
const router = express.Router();
const config = require('../config/config');
const { Loo } = require('../db')(config.mongo.url);

/**
 * Get loos near lon/lat
 * Accepts `radius` in meters in query
 */
router.get('/near/:lon/:lat', async (req, res) => {
  const wantedRadius =
    parseFloat(req.query.radius) || config.query_defaults.defaultRadius;
  const allowedRadius = Math.min(wantedRadius, config.query_defaults.maxRadius);
  const loos = await Loo.findNear(
    parseFloat(req.params.lon),
    parseFloat(req.params.lat),
    allowedRadius
  ).exec();

  res.status(200).json(loos);
});

/**
 * GET an array of points
 */
router.get('/', async (req, res) => {
  const loos = await Loo.find({}, 'type geometry').exec();
  res.status(200).json(loos);
});

/**
 * Regenerate a loo from its reports
 * this'll update it's admin geo
 */
router.get('/:id/updateArea', async (req, res) => {
  const loo = await Loo.findById(req.params.id);
  const regen = await loo.regenerate();
  await regen.save();
  res.status(200).json(regen);
});

/**
 * GET a loo by ID
 */
router.get('/:id', async (req, res) => {
  const loo = await Loo.findById(req.params.id).exec();
  console.log(loo);
  if (!loo) {
    return res.status(404).end();
  }

  if (req.query.populateReports) {
    const populatedReportsLoo = await loo.populate('reports').execPopulate();
    res.status(200).json(populatedReportsLoo);
    return;
  }

  res.status(200).json(loo);
});

module.exports = router;
