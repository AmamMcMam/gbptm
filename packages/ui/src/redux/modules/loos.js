export const FIND_NEARBY_REQUEST = 'LOOS/FIND_NEARBY/REQUEST';
export const FIND_NEARBY_SUCCESS = 'LOOS/FIND_NEARBY/SUCCESS';
export const FIND_BY_ID_REQUEST = 'LOOS/FIND_BY_ID_REQUEST/SUCCESS';
export const FIND_BY_ID_SUCCESS = 'LOOS/FIND_BY_ID_SUCCESS/SUCCESS';
export const REPORT_REQUEST = 'LOOS/REPORT/REQUEST';
export const REPORT_SUCCESS = 'LOOS/REPORT/SUCCESS';
export const REMOVE_REQUEST = 'LOOS/REMOVE/REQUEST';
export const REMOVE_SUCCESS = 'LOOS/REMOVE/SUCCESS';
export const REPORT_PROCESS_PENDING = 'LOOS/REPORT_PROCESS_PENDING';

// Actions

export const actionFindNearbyRequest = (lng, lat, radius) => ({
  type: FIND_NEARBY_REQUEST,
  payload: {
    lng,
    lat,
    radius,
  },
});

export const actionFindNearbySuccess = loos => ({
  type: FIND_NEARBY_SUCCESS,
  payload: {
    loos,
  },
});

export const actionFindByIdRequest = id => ({
  type: FIND_BY_ID_REQUEST,
  payload: {
    id,
  },
});

export const actionFindByIdSuccess = loo => ({
  type: FIND_BY_ID_SUCCESS,
  payload: {
    loo,
  },
});

export const actionReportRequest = (loo, from) => ({
  type: REPORT_REQUEST,
  payload: {
    loo,
  },
});

export const actionReportSuccess = loo => ({
  type: REPORT_SUCCESS,
  payload: loo,
});

export const actionRemoveRequest = (looId, reason) => ({
  type: REMOVE_REQUEST,
  payload: {
    looId,
    reason,
  },
});

export const actionRemoveSuccess = () => ({
  type: REMOVE_SUCCESS,
});

export const actionProcessPendingReports = () => ({
  type: REPORT_PROCESS_PENDING,
});

export const actions = {
  [FIND_NEARBY_REQUEST]: actionFindNearbyRequest,
  [FIND_NEARBY_SUCCESS]: actionFindNearbySuccess,
  [FIND_BY_ID_REQUEST]: actionFindByIdRequest,
  [FIND_BY_ID_SUCCESS]: actionFindByIdSuccess,
  [REPORT_REQUEST]: actionReportRequest,
  [REPORT_SUCCESS]: actionReportSuccess,
  [REMOVE_REQUEST]: actionRemoveRequest,
  [REMOVE_SUCCESS]: actionRemoveSuccess,
  [REPORT_PROCESS_PENDING]: actionProcessPendingReports,
};

// Action Handlers

const ACTION_HANDLERS = {
  [FIND_BY_ID_REQUEST]: function(state, action) {
    // Clear the stored loo reference on each request to ensure we
    // don't render against a cached instance
    return {
      ...state,
      byId: null,
    };
  },

  [FIND_NEARBY_SUCCESS]: function(state, action) {
    return {
      ...state,
      nearby: action.payload.loos || [],
    };
  },

  [FIND_BY_ID_SUCCESS]: function(state, action) {
    return {
      ...state,
      byId: action.payload.loo,
    };
  },
};

// Reducer
export default function loosReducer(state = {}, action) {
  const handler = ACTION_HANDLERS[action.type];
  return handler ? handler(state, action) : state;
}
