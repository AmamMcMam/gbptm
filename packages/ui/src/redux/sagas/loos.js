import { all, takeLatest, call, put } from 'redux-saga/effects';
import history from '../../history';

import api from '../../api';
import { PENDING_REPORT_KEY, PENDING_REMOVE_KEY } from '../../config';

import {
  FIND_NEARBY_REQUEST,
  FIND_BY_ID_REQUEST,
  REPORT_REQUEST,
  REMOVE_REQUEST,
  actionFindNearbySuccess,
  actionFindByIdSuccess,
  actionReportSuccess,
  actionRemoveSuccess,
} from '../modules/loos';

import { LOGGED_IN } from '../modules/auth';

export default function makeLoosSaga(auth) {
  function* findNearbyLoosSaga(action) {
    var { lng, lat, radius } = action.payload;
    var loos = yield call(api.findLoos, lng, lat, radius);
    yield put(actionFindNearbySuccess(loos));
  }
  function* findLooByIdSaga(action) {
    var loo = yield call(api.findLooById, action.payload.id);
    yield put(actionFindByIdSuccess(loo));
  }

  function* report(loo) {
    // Todo: Catch HTTP 401 and navigate to '/login'
    const ids = yield call(api.reportLoo, loo, auth.getAccessToken());
    // maybe we should navigate as a result of the success action
    yield put(actionReportSuccess(ids));
    return yield call(history.push, `/loos/${ids.loo}`);
  }

  function* reportRequest(action) {
    // TODO: when a report has an existing loo dispatch `actionUncacheById`
    const loo = action.payload.loo;
    // Re-direct the user to the authentication screen, if they're not
    // already logged in
    if (!auth.isAuthenticated()) {
      // Store loo in localStorage so, if neccessary, we can perform
      // authentication and pick it up again once we've been re-directed
      // back to our application
      localStorage.setItem(PENDING_REPORT_KEY, JSON.stringify(loo));
      return yield call(history.push, '/login');
    }
    yield call(report, loo);
  }

  function* remove(id, reason) {
    // Todo: Catch HTTP 401 and navigate to '/login'
    const result = yield call(api.removeLoo, id, reason, auth.getAccessToken());
    // maybe we should navigate as a result of the success action
    yield put(actionRemoveSuccess(result));
    return yield call(history.push, `/`);
  }

  function* removeRequest(action) {
    const { looId, reason } = action.payload;
    // Re-direct the user to the authentication screen, if they're not
    // already logged in
    if (!auth.isAuthenticated()) {
      // Store loo in localStorage so, if neccessary, we can perform
      // authentication and pick it up again once we've been re-directed
      // back to our application
      localStorage.setItem(PENDING_REMOVE_KEY, JSON.stringify(action.payload));
      return yield call(history.push, '/login');
    }
    yield call(remove, looId, reason);
  }

  /**
   * Called on login to pick up any pending reports
   */
  function* processPending() {
    const loo = JSON.parse(localStorage.getItem(PENDING_REPORT_KEY));
    if (loo) {
      localStorage.removeItem(PENDING_REPORT_KEY);
      yield call(report, loo);
    }
    const remove = JSON.parse(localStorage.getItem(PENDING_REMOVE_KEY));
    if (remove) {
      localStorage.removeItem(PENDING_REMOVE_KEY);
      yield call(report, loo);
    }
  }

  return function* loosSaga() {
    yield all([
      takeLatest(FIND_NEARBY_REQUEST, findNearbyLoosSaga),
      takeLatest(FIND_BY_ID_REQUEST, findLooByIdSaga),
      takeLatest(REPORT_REQUEST, reportRequest),
      takeLatest(REMOVE_REQUEST, removeRequest),
      takeLatest(LOGGED_IN, processPending),
    ]);
  };
}
