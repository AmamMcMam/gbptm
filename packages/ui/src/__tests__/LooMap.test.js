import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import {
  LooMap,
  GeolocationMapControl,
  LocateMapControl,
} from '../components/map/LooMap';

import loos from './fixtures/loos.json';

jest.mock('leaflet');

describe('LooMap', () => {
  var defaultProps = {
    loos,
    initialPosition: {
      lat: 1.292708,
      lng: 52.6369559,
    },
  };

  // Jest snapshot
  it('renders snapshot', () => {
    var tree = renderer.create(<LooMap {...defaultProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  // Todo: Test mounted component to ensure `leafletElement` properties are set correctly.

  describe('Controls: ', () => {
    it('render', () => {
      var component = shallow(
        <LooMap
          {...defaultProps}
          showSearchControl={true}
          showLocateControl={true}
        />
      );

      expect(component.find(GeolocationMapControl).length).toBe(1);
      expect(component.find(LocateMapControl).length).toBe(1);
    });

    it('hide', () => {
      var component = shallow(
        <LooMap
          {...defaultProps}
          showSearchControl={false}
          showLocateControl={false}
        />
      );

      expect(component.find(GeolocationMapControl).length).toBe(0);
      expect(component.find(LocateMapControl).length).toBe(0);
    });
  });
});
