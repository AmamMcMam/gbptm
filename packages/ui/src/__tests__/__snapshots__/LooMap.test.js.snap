// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LooMap renders snapshot 1`] = `
<div
  className="map"
  id={undefined}
  style={undefined}
/>
`;
